var/spack/repos/builtin/packages/hdf5/package.py

```diff
@@ -223,106 +230,100 @@ def libs(self):
-            # files, the second one points to the directory with library files.
-            # If the second path is not specified, the configure script assumes
-            # that it equals to prefix/lib. However, the correct directory
-            # might be prefix/lib64. It is not a problem when the building is
-            # done with Spack's compiler wrapper but it makes the Libtool
-            # files (*.la) invalid, which makes it problematic to use the
-            # installed library outside of Spack environment.
-            extra_args.append('--with-szlib=%s,%s' %
-                              (szip_spec.headers.directories[0],
-                               szip_spec.libs.directories[0]))
+        args.append('-DALLOW_UNSUPPORTED=ON')
+
+        spec = self.spec
+        if '+zlib' in spec:
+            args.append('-DHDF5_ENABLE_Z_LIB_SUPPORT:BOOL=ON')
+            args.append(
+                '-DZLIB_INCLUDE_DIR:PATH={0}'.format(
+                    spec['zlib'].prefix.include))
+            args.append(
+                '-DZLIB_DIR:PATH={0}'.format(
+                    spec['zlib'].prefix.lib))
         else:
-            extra_args.append('--without-szlib')
+            args.append('-DHDF5_ENABLE_Z_LIB_SUPPORT:BOOL=OFF')
+
+        if '+mpi' in spec:
+            args.append('-DHDF5_ENABLE_PARALLEL=ON')
+            args.append('-DCMAKE_C_COMPILER={0}'.format(spec['mpi'].mpicc))
+            args.append('-DCMAKE_CXX_COMPILER={0}'.format(spec['mpi'].mpicxx))
+            args.append(
+                '-DCMAKE_Fortran_COMPILER={0}'.format(spec['mpi'].mpifc))
+        else:
+            args.append('-DHDF5_ENABLE_PARALLEL=OFF')
 
-        if self.spec.satisfies('@1.10:'):
-            if '+debug' in self.spec:
-                extra_args.append('--enable-build-mode=debug')
-            else:
-                extra_args.append('--enable-build-mode=production')
+        if '+static' in spec:
+            args.append('-DONLY_SHARED_LIBS=OFF')
         else:
-            if '+debug' in self.spec:
-                extra_args.append('--enable-debug=all')
-            else:
-                extra_args.append('--enable-production')
+            args.append('-DONLY_SHARED_LIBS=ON')
 
-            # '--enable-fortran2003' no longer exists as of version 1.10.0
-            if '+fortran' in self.spec:
-                extra_args.append('--enable-fortran2003')
-            else:
-                extra_args.append('--disable-fortran2003')
+        if '+debug' in spec:
+            args.append('-DCMAKE_BUILD_TYPE=Debug')
+        else:
+            args.append('-DCMAKE_BUILD_TYPE=Release')
 
-        if '+shared' in self.spec:
-            extra_args.append('--enable-shared')
+        if '+threadsafe' in spec:
+            args.append('-DHDF5_ENABLE_THREADSAFE=ON')
         else:
-            extra_args.append('--disable-shared')
-            extra_args.append('--enable-static-exec')
-
-        if '+pic' in self.spec:
```

Comment:
It's also important to make sure that static libraries are compiled with pic flags when `+pic`.

Reply:
That's not always the case, especially if using CrayPE wrappers in static only mode, in which case PIC isn't even really supported.

Reply:
Missed this one - need to get more info if it shouldn't always be on by default.

Reply:
There's been a multi-year discussion about `static`/`static+pic`/`shared`/`shared+static` variants... @alalazo

Reply:
With CMake, HDF5 tells CMake to use the PIC flag if it can with "set (CMAKE_POSITION_INDEPENDENT_CODE ON)" beginning in version 1.8.13, so it should always be added if available.  Versions 1.8.10 and 1.3.12 added PIC for all platforms except Windows.  We don't believe the variant is necessary since CMake will add the flag if it can.

Reply:
Change of mind after further discussion:  we'll keep the variant and use it to set the CMake variable 'CMAKE_POSITION_INDEPENDENT_CODE' to True or False according to the 'pic' variant value with " self.define_from_variant('CMAKE_POSITION_INDEPENDENT_CODE', 'pic')".  CMake will set a PIC flag when available when +pic is in the spec, but not for ~pic. The default for 'pic' is still True.

Reply:
@lrknox I would like to comment on why we currently have the variant `pic` and inject PIC flags if it is set:
1. We want to be able to disable the flag injection with `~pic` if, for whatever reason, it breaks the building. This is a way for the users to tell Spack not to try to be smarter than them or the building system of `hdf5`.
2. Libtool (unless patched like in `parallel-netcdf`) cannot always recognize the real compilers behind the MPI wrappers (Libtool relies on the basename of the compiler executable in most cases) and, therefore, is unable to set the correct PIC flag. The shared versions of the libraries simply cannot be built in such cases, so we have to set the flags explicitly. As far as I know, this should not be a problem for CMake.
3. Libtool is old and might not know the correct PIC flag for the compiler in use. Again, we have to set the flags. This partially holds for CMake: its older versions might not know the correct flag for the most recent compilers.
4. By default, and when both `--enable-shared` and `--enable-static` configure options are set, Libtool compiles each file twice: with the PIC flag for the shared library and without the flag for the static one. This is supposed to give the users of the library a choice: benefit from the features of shared libraries or use the static library and gain some performance because position-dependent code (i.e. non-PIC) is supposed to be slightly faster. This, however, prevents the users from building a shared version of some dependent library and link it to the static version of `hdf5` (I don't really know why this might be needed by I remember people asking for this).

Whatever you decide here, please, make sure that those issues are addresses.

Reply:
Status as of 4/29 changes
1. ~pic will not add flags for +pic in this package.py file, and will also set CMAKE_POSITION_INDEPENDENT_CODE to False which should stop CMake from adding any PIC flags.
2, 3. The same pic flags that were added in the spack develop version were kept here for +pic which is the default. They should overwrite any flags that CMake would set.
4. In response to previous discussion, there is no static variant.  Shared libs are produced only if +shared and static libs only if ~shared.  I suppose you can build them one at a time with +pic or ~pic?